openlr/router.cpp

```diff
@@ -0,0 +1,789 @@
+  if (m_stageStartDistance != rhs.m_stageStartDistance)
+    return m_stageStartDistance < rhs.m_stageStartDistance;
+  if (m_stage != rhs.m_stage)
+    return m_stage < rhs.m_stage;
+  return m_bearingChecked < rhs.m_bearingChecked;
+}
+
+bool Router::Vertex::operator==(Vertex const & rhs) const
+{
+  return m_junction == rhs.m_junction && m_stageStart == rhs.m_stageStart &&
+         m_stageStartDistance == rhs.m_stageStartDistance && m_stage == rhs.m_stage &&
+         m_bearingChecked == rhs.m_bearingChecked;
+}
+
+// Router::Edge ------------------------------------------------------------------------------------
+Router::Edge::Edge(Vertex const & u, Vertex const & v, routing::Edge const & raw, bool isSpecial)
+  : m_u(u), m_v(v), m_raw(raw), m_isSpecial(isSpecial)
+{
+}
+
+// static
+Router::Edge Router::Edge::MakeNormal(Vertex const & u, Vertex const & v,
+                                             routing::Edge const & raw)
+{
+  return Edge(u, v, raw, false /* isSpecial */);
+}
+
+// static
+Router::Edge Router::Edge::MakeSpecial(Vertex const & u, Vertex const & v)
+{
+  return Edge(u, v, routing::Edge::MakeFake(u.m_junction, v.m_junction, false /* partOfReal */),
+              true /* isSpecial */);
+}
+
+pair<m2::PointD, m2::PointD> Router::Edge::ToPair() const
+{
+  auto const & e = m_raw;
+  return make_pair(e.GetStartJunction().GetPoint(), e.GetEndJunction().GetPoint());
+}
+
+pair<m2::PointD, m2::PointD> Router::Edge::ToPairRev() const
+{
+  auto const & e = m_raw;
+  return make_pair(e.GetEndJunction().GetPoint(), e.GetStartJunction().GetPoint());
+}
+
+// Router::Router ----------------------------------------------------------------------------------
+Router::Router(routing::FeaturesRoadGraph & graph, RoadInfoGetter & roadInfoGetter)
+  : m_graph(graph), m_roadInfoGetter(roadInfoGetter)
+{
+}
+
+bool Router::Go(vector<WayPoint> const & points, double positiveOffsetM, double negativeOffsetM,
+                vector<routing::Edge> & path)
+{
+  if (!Init(points, positiveOffsetM, negativeOffsetM))
+    return false;
+  return FindPath(path);
+}
+
+bool Router::Init(vector<WayPoint> const & points, double positiveOffsetM, double negativeOffsetM)
+{
+  CHECK_GREATER_OR_EQUAL(points.size(), 2, ());
+
+  m_points = points;
+  m_positiveOffsetM = positiveOffsetM;
+  m_negativeOffsetM = negativeOffsetM;
+
+  m_graph.ResetFakes();
+
+  m_pivots.clear();
+  for (size_t i = 1; i + 1 < m_points.size(); ++i)
+  {
+    m_pivots.emplace_back();
+    auto & ps = m_pivots.back();
+
+    vector<pair<routing::Edge, routing::Junction>> vicinity;
+    m_graph.FindClosestEdges(m_points[i].m_point, kMaxRoadCandidates, vicinity);
+    for (auto const & v : vicinity)
+    {
+      auto const & e = v.first;
+      ps.push_back(e.GetStartJunction().GetPoint());
+      ps.push_back(e.GetEndJunction().GetPoint());
+    }
+
+    if (ps.empty())
+      return false;
+  }
+
+  m_pivots.push_back({m_points.back().m_point});
```

Comment:
Please, take a look at the top of the method: CHECK_GREATER_OR_EQUAL(points.size(), 2, ());